// @google-cloud/trace-agent/build/src/util.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
// Copyright 2015 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
const path = require("path");
const sourceMapSupport = require("source-map-support");
const { hexToDec, decToHex, } = require('hex2dec');
exports.hexToDec = hexToDec;
exports.decToHex = decToHex;
// This symbol must be exported (for now).
// See: https://github.com/Microsoft/TypeScript/issues/20080
exports.kSingleton = Symbol();
exports.FORCE_NEW = Symbol('force-new');
/**
 * A class that provides access to a singleton.
 * We assume that any such singleton is always constructed with two arguments:
 * An arbitrary configuration object and a logger.
 * Instances of this type should only be constructed in module scope.
 */
class Singleton {
    constructor(implementation) {
        this.implementation = implementation;
        // Note: private[symbol] is enforced by clang-format.
        this[_a] = null;
    }
    create(config, logger) {
        if (!this[exports.kSingleton] || config[exports.FORCE_NEW]) {
            const s = this[exports.kSingleton];
            if (s && s.disable) {
                s.disable();
            }
            this[exports.kSingleton] = new this.implementation(config, logger);
            return this[exports.kSingleton];
        }
        else {
            throw new Error(`${this.implementation.name} has already been created.`);
        }
    }
    get() {
        if (!this[exports.kSingleton]) {
            throw new Error(`${this.implementation.name} has not yet been created.`);
        }
        return this[exports.kSingleton];
    }
    exists() {
        return !!this[exports.kSingleton];
    }
}
exports.Singleton = Singleton;
_a = exports.kSingleton;
/**
 * Returns the last parameter that is not null, undefined, or NaN.
 * @param defaultValue The first parameter. This must not be null/undefined/NaN.
 * @param otherValues Other parameters, which may be null/undefined/NaN.
 */
function lastOf(defaultValue, ...otherValues) {
    for (let i = otherValues.length - 1; i >= 0; i--) {
        // tslint:disable:no-any
        if (otherValues[i] !== null &&
            otherValues[i] !== undefined &&
            (typeof otherValues[i] !== 'number' || !isNaN(otherValues[i]))) {
            return otherValues[i];
        }
        // tslint:enable:no-any
    }
    return defaultValue;
}
exports.lastOf = lastOf;
/**
 * Truncates the provided `string` to be at most `length` bytes
 * after utf8 encoding and the appending of '...'.
 * We produce the result by iterating over input characters to
 * avoid truncating the string potentially producing partial unicode
 * characters at the end.
 */
function truncate(str, length) {
    if (Buffer.byteLength(str, 'utf8') <= length) {
        return str;
    }
    str = str.substr(0, length - 3);
    while (Buffer.byteLength(str, 'utf8') > length - 3) {
        str = str.substr(0, str.length - 1);
    }
    return str + '...';
}
exports.truncate = truncate;
// Includes support for npm '@org/name' packages
// Regex: .*?node_modules(?!.*node_modules)\/(@[^\/]*\/[^\/]*|[^\/]*).*
// Tests: https://regex101.com/r/lW2bE3/6
const moduleRegex = new RegExp([
    '.*?node_modules(?!.*node_modules)\\',
    '(@[^\\',
    ']*\\',
    '[^\\',
    ']*|[^\\',
    ']*).*',
].join(path.sep));
/**
 * Parse a cookie-style header string to extract traceId, spandId and options
 * ex: '123456/667;o=3'
 * -> {traceId: '123456', spanId: '667', options: '3'}
 * note that we ignore trailing garbage if there is more than one '='
 * Returns null if traceId or spanId could not be found.
 *
 * @param str string representation of the trace headers
 * @return object with keys. null if there is a problem.
 */
function parseContextFromHeader(str) {
    if (!str) {
        return null;
    }
    const matches = str.match(/^([0-9a-fA-F]+)(?:\/([0-9]+))(?:;o=(.*))?/);
    if (!matches ||
        matches.length !== 4 ||
        matches[0] !== str ||
        (matches[2] && isNaN(Number(matches[2])))) {
        return null;
    }
    return {
        traceId: matches[1],
        spanId: matches[2],
        options: isNaN(Number(matches[3])) ? undefined : Number(matches[3]),
    };
}
exports.parseContextFromHeader = parseContextFromHeader;
/**
 * Generates a trace context header value that can be used
 * to follow the associated request through other Google services.
 *
 * @param traceContext An object with information sufficient for creating a
 *        serialized trace context.
 */
function generateTraceContext(traceContext) {
    if (!traceContext) {
        return '';
    }
    let header = `${traceContext.traceId}/${traceContext.spanId}`;
    if (typeof traceContext.options !== 'undefined') {
        header += `;o=${traceContext.options}`;
    }
    return header;
}
exports.generateTraceContext = generateTraceContext;
/**
 * Retrieves a package name from the full import path.
 * For example:
 *   './node_modules/bar/index/foo.js' => 'bar'
 *
 * @param path The full import path.
 */
function packageNameFromPath(importPath) {
    const matches = moduleRegex.exec(importPath);
    return matches && matches.length > 1 ? matches[1] : null;
}
exports.packageNameFromPath = packageNameFromPath;
/**
 * Creates a StackFrame object containing a structured stack trace.
 * @param numFrames The number of frames to retain.
 * @param skipFrames The number of top-most frames to remove.
 * @param constructorOpt A function passed to Error.captureStackTrace, which
 *   causes it to ignore the frames above the top-most call to this function.
 */
function createStackTrace(numFrames, skipFrames, constructorOpt) {
    // This is a mechanism to get the structured stack trace out of V8.
    // prepareStackTrace is called the first time the Error#stack property is
    // accessed. The original behavior is to format the stack as an exception
    // throw, which is not what we like. We customize it.
    //
    // See: https://code.google.com/p/v8-wiki/wiki/JavaScriptStackTraceApi
    //
    if (numFrames === 0) {
        return [];
    }
    const origLimit = Error.stackTraceLimit;
    Error.stackTraceLimit = numFrames + skipFrames;
    const origPrepare = Error.prepareStackTrace;
    Error.prepareStackTrace = (error, structured) => {
        return structured.map(sourceMapSupport.wrapCallSite);
    };
    const e = {};
    Error.captureStackTrace(e, constructorOpt);
    const stackFrames = [];
    if (e.stack) {
        e.stack.forEach((callSite, i) => {
            if (i < skipFrames) {
                return;
            }
            // TODO(kjin): Check if callSite getters actually return null or
            // undefined. Docs say undefined but we guard it here just in case.
            const functionName = callSite.getFunctionName();
            const methodName = callSite.getMethodName();
            const name = methodName && functionName
                ? functionName + ' [as ' + methodName + ']'
                : functionName || methodName || '<anonymous function>';
            const stackFrame = {
                method_name: name,
                file_name: callSite.getFileName() || undefined,
                line_number: callSite.getLineNumber() || undefined,
                column_number: callSite.getColumnNumber() || undefined,
            };
            stackFrames.push(stackFrame);
        });
    }
    Error.stackTraceLimit = origLimit;
    Error.prepareStackTrace = origPrepare;
    return stackFrames;
}
exports.createStackTrace = createStackTrace;
/**
 * Serialize the given trace context into a Buffer.
 * @param traceContext The trace context to serialize.
 */
function serializeTraceContext(traceContext) {
    //  0           1           2
    //  0 1 2345678901234567 8 90123456 7 8
    // -------------------------------------
    // | | |                | |        | | |
    // -------------------------------------
    //  ^ ^      ^           ^    ^     ^ ^
    //  | |      |           |    |     | `-- options value (traceContext.options)
    //  | |      |           |    |     `---- options field ID (2)
    //  | |      |           |    `---------- spanID value (traceConext.spanID)
    //  | |      |           `--------------- spanID field ID (1)
    //  | |      `--------------------------- traceID value (traceContext.traceID)
    //  | `---------------------------------- traceID field ID (0)
    //  `------------------------------------ version (0)
    const result = Buffer.alloc(29, 0);
    result.write(traceContext.traceId, 2, 16, 'hex');
    result.writeUInt8(1, 18);
    // Convert Span ID from decimal to base 16 representation, then left pad into
    // a length-16 hex string. (decToHex prepends its output with '0x', so we
    // also slice that off.)
    const base16SpanId = `0000000000000000${decToHex(traceContext.spanId).slice(2)}`.slice(-16);
    result.write(base16SpanId, 19, 8, 'hex');
    result.writeUInt8(2, 27);
    result.writeUInt8(traceContext.options || 0, 28);
    return result;
}
exports.serializeTraceContext = serializeTraceContext;
/**
 * Deseralize the given trace context from binary encoding. If the input is a
 * Buffer of incorrect size or unexpected fields, then this function will return
 * null.
 * @param buffer The trace context to deserialize.
 */
function deserializeTraceContext(buffer) {
    const result = { traceId: '', spanId: '' };
    // Length must be 29.
    if (buffer.length !== 29) {
        return null;
    }
    // Check version and field numbers.
    if (buffer.readUInt8(0) !== 0 ||
        buffer.readUInt8(1) !== 0 ||
        buffer.readUInt8(18) !== 1 ||
        buffer.readUInt8(27) !== 2) {
        return null;
    }
    // See serializeTraceContext for byte offsets.
    result.traceId = buffer.slice(2, 18).toString('hex');
    result.spanId = hexToDec(buffer.slice(19, 27).toString('hex'));
    result.options = buffer.readUInt8(28);
    return result;
}
exports.deserializeTraceContext = deserializeTraceContext;
